src/main/scala/pipedsl/typechecker/TypeInferenceWrapper.scala

```diff
@@ -80,50 +80,62 @@ object TypeInferenceWrapper
    }
   }
 
-  private def type_subst_map_fopt(t :Type, tp_mp: mutable.HashMap[Id, Type]) :Option[Type] = try
+  private def type_subst_map_fopt(t :Type, tp_mp: mutable.HashMap[Id, Type], templated :mutable.HashSet[Id] = mutable.HashSet.empty) :Option[Type] = try
     {
-     Some(type_subst_map(t, tp_mp))
+     Some(type_subst_map(t, tp_mp, templated))
     } catch
    {
     case IntWidthNotSpecified() => None
    }
 
-  private def type_subst_map(t: Type, tp_mp: mutable.HashMap[Id, Type]): Type = t match
-  {
-   case TSignVar(nm) => tp_mp.get(nm) match
-   {
-    case Some(value) => type_subst_map(value, tp_mp)
-   }
-   case sz@TSizedInt(len, sign) =>
-    val width = type_subst_map(len, tp_mp).copyMeta(sz) |> to_width
-    val sn = type_subst_map(sign, tp_mp).copyMeta(sign) |> to_sign
-    sz.copy(len = width, sign = sn)
-   case f@TFun(args, ret) => f.copy(args = args.map(type_subst_map(_, tp_mp)), ret = type_subst_map(ret, tp_mp)).copyMeta(f)
-   case r@TRecType(_, fields) => r.copy(fields = fields.map(idtp => (idtp._1, type_subst_map(idtp._2, tp_mp)))).copyMeta(r)
-   case m: TMemType => m.copy(elem = type_subst_map(m.elem, tp_mp)).copyMeta(m)
-   case m@TModType(inputs, refs, _, _) => m.copy(inputs = inputs.map(type_subst_map(_, tp_mp)), refs = refs.map(type_subst_map(_, tp_mp))).copyMeta(m)
-   case l@TLockedMemType(mem, _, _) => l.copy(mem = type_subst_map(mem, tp_mp).asInstanceOf[TMemType]).copyMeta(l)
-   case TNamedType(name) => tp_mp.get(name) match
-   {
-    case Some(value) => type_subst_map(value, tp_mp)
-    case None => throw IntWidthNotSpecified()
-   }
-   case m@TMaybe(btyp) => m.copy(btyp = type_subst_map(btyp, tp_mp))
-   case TBitWidthAdd(b1, b2) =>
-    val len1 = type_subst_map(b1, tp_mp) |> to_len
-    val len2 = type_subst_map(b2, tp_mp) |> to_len
-    TBitWidthLen(len1 + len2)
-   case TBitWidthMax(b1, b2) =>
-    val len1 = type_subst_map(b1, tp_mp) |> to_len
-    val len2 = type_subst_map(b2, tp_mp) |> to_len
-    TBitWidthLen(Math.max(len1, len2))
-   case TBitWidthVar(name) => tp_mp.get(name) match
+  private def type_subst_map(t: Type, tp_mp: mutable.HashMap[Id, Type], templated :mutable.HashSet[Id]): Type =
```

Comment:
We're gonna need to walk through this one together haha, it's a doozy of a change.

Reply:
ah yes, magic